velox/expression/Expr.cpp

```diff
@@ -975,6 +1005,72 @@ void Expr::evalAll(
   if (remainingRows != &rows) {
     addNulls(rows, remainingRows->asRange().bits(), context, result);
   }
+  context.releaseVectors(inputValues_);
+  inputValues_.clear();
+}
+
+void Expr::evalFlatNonNull(
+    const SelectivityVector& rows,
+    EvalCtx& context,
+    VectorPtr& result) {
+  if (isMultiplyReferenced_ && !inputs_.empty()) {
+    evalSharedSubexpr(rows, context, result);
+    return;
+  }
+  ExceptionContextSetter exceptionContext(
+      {[](auto* expr) { return static_cast<Expr*>(expr)->toString(); }, this});
+
+  if (isSpecialForm()) {
+    evalSpecialForm(rows, context, result);
+    return;
+  }
+  LocalSelectivityVector nonNulls(context);
+  auto* remainingRows = &rows;
+
+  inputValues_.resize(inputs_.size());
+  bool tryPeelArgs = deterministic_;
+  bool anyUnique = false;
+  for (int32_t i = 0; i < inputs_.size(); ++i) {
+    inputs_[i]->evalFlatNonNull(*remainingRows, context, inputValues_[i]);
```

Comment:
Even though none of the inputs are NULL, it's still possible for a UDF to produce NULL.  I'm having trouble thinking of an example, but there's nothing in the interface that prevents it.

Are you assuming here that this doesn't happen, or that it happens so rarely that the performance cost of checking for NULLs doesn't outweigh the benefit of skipping the evaluation of rows with NULLs?